homeassistant/components/fujitsu_hvac/climate.py

```diff
@@ -0,0 +1,202 @@
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import (
+    API_TIMEOUT,
+    AYLA_APP_ID,
+    AYLA_APP_SECRET,
+    CONF_DEVICE,
+    CONF_EUROPE,
+    DOMAIN,
+    FAN_MODE_MAP,
+    HVAC_MODE_MAP,
+    SWING_MODE_MAP,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up one Fujitsu HVAC device."""
+    api = new_ayla_api(
+        entry.data[CONF_USERNAME],
+        entry.data[CONF_PASSWORD],
+        AYLA_APP_ID,
+        AYLA_APP_SECRET,
+        europe=entry.data[CONF_EUROPE],
+    )
+
+    try:
+        async with timeout(API_TIMEOUT):
+            await api.async_sign_in()
+    except (TimeoutError, AylaAuthError):
+        return
+
+    devices = await api.async_get_devices()
+    for dev in devices:
+        if dev.device_serial_number == entry.data[CONF_DEVICE]:
+            await dev.async_update()
+            async_add_entities([FujitsuHVACDevice(dev)])
+            return
+
+
+class FujitsuHVACDevice(ClimateEntity):
+    """Represent a Fujitsu HVAC device."""
+
+    _attr_temperature_unit = UnitOfTemperature.CELSIUS
+    _attr_precision = 0.5
+    _attr_target_temperature_step = 0.5
+    _attr_has_entity_name = True
+    _attr_name = None
+
+    def __init__(self, dev: FujitsuHVAC) -> None:
+        """Set up static attributes."""
+        self._dev = dev
+
+        self._attr_unique_id = f"{DOMAIN}_{self._dev.device_serial_number}"
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, self._dev.device_serial_number)},
+            name=self._dev.device_name,
+            manufacturer="Fujitsu",
+            model=self._dev.property_values["model_name"],
+            serial_number=self._dev.device_serial_number,
+            sw_version=self._dev.property_values["mcu_firmware_version"],
+        )
+        self.update_attributes()
+
+    async def async_update(self) -> None:
+        """Gather data from API and update our attributes."""
+        try:
+            if self._dev.ayla_api.token_expiring_soon:
+                await self._dev.ayla_api.refresh_auth()
+        except AylaAuthError:
+            await self._dev.ayla_api.async_sign_in()
+
+        await self._dev.async_update()
+        self.update_attributes()
+
+    def update_attributes(self) -> None:
+        """Update all dynamic attributes."""
+        self._attr_supported_features = self._get_supported_features()
+        self._attr_current_temperature = self._dev.sensed_temp
+        self._attr_fan_mode = self._get_fan_mode()
+        self._attr_fan_modes = self._get_fan_modes()
+        self._attr_hvac_mode = self._get_hvac_mode()
+        self._attr_hvac_modes = self._get_hvac_modes()
+        self._attr_swing_mode = self._get_swing_mode()
+        self._attr_swing_modes = self._get_swing_modes()
+        self._attr_target_temperature = self._get_target_temperature()
```

Comment:
imo it'd be cleaner if we used the properties here instead of the attributes, because then we only need one "function"

Reply:
I'm not sure I understand. Specifically the part about one "function".

Reply:
I agree with @zweckj, you might as well transform the functions to properties.

```py
def _get_fan_mode(self) -> str | None:
```

```py
@property
def fan_mode(self) -> str | None:
```

Please checkout other integrations which use this (there are many, don't have an example right now (I think flexit_bacnet uses it this way))